File: Py-Solutions/Problem-12.py
# What is the value of the first triangle number to have over five hundred divisors?
import math

def numDivisors(num):
  divs = 0
  for i in range(1,1 + int(math.floor(math.sqrt(num)))):
    if(num % i == 0):
      divs += 1
  if math.sqrt(num) % 1 == 0:
    return divs * 2 - 1
  return divs * 2

def triDivisors(num_divisors):
  tri_num = 1
  adder = 2
  while numDivisors(tri_num) <= num_divisors:
    tri_num += adder
    adder += 1
  return [tri_num, numDivisors(tri_num)]

def main():
  print(triDivisors(500))

if __name__ == '__main__':
  main()

File: JS-Solutions/Problem-12.js
// What is the value of the first triangle number to have over five hundred divisors?

const triMaker = (len) => {
  let num = 0;
  let adder = 1;
  while (adder <= len) {
    num += adder;
    adder += 1;
  }
  return num
}

const numDivisors = (num) => {
  let total = 1;
  for (let i=2;i<Math.round(num**0.5) + 1;i++) {
    if (num % i == 0) {
      total += 1;
    }
  }
  total *= 2;
  if (num**0.5 % 1 == 0) {
    total -= 1;
  }
  return total
}

const longDivisorList = (len) => {
  let max_divisors = 0;
  let n = 1;
  let tri_num, divs;
  while (max_divisors <= len) {
    tri_num = triMaker(n);
    divs = numDivisors(tri_num);
    if (divs > max_divisors) {
      max_divisors = divs;
    }
    n += 1;
  }
  return [tri_num, divs];
}

console.log(longDivisorList(500));
